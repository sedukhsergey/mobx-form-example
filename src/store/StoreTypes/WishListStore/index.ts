export type WishListItem = {
  name: string,
  price: number,
  image: string,
  changeName: (item: string) => void,
  changePrice: (item: number) => void,
  changeImage: (item: string) => void,
  remove: () => void,
}

export type WishListStore = {
  items: WishListItem[],
  add: (item: WishListItem) => void,
  createItems: WishListItem[],
  fetchItems: () => void,
  remove: WishListItem,
  totalPrice: () => void,
}
