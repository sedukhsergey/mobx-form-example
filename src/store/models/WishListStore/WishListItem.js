import {
  types, getParent
} from 'mobx-state-tree';

const WishListItem = types
  .model({
    name: types.string,
    price: types.number,
    image: types.optional(types.string, ''),
  })
  .actions(self => ({
    changeName(newName) {
      self.name = newName;
    },
    changePrice(newPrice) {
      self.price = newPrice;
    },
    changeImage(newImage) {
      self.image = newImage;
    },
    remove() {
      getParent(self, 2).remove(self);
    },
  }));

export default WishListItem;
