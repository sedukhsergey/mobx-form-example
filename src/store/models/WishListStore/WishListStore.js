import {
  types, destroy
} from 'mobx-state-tree';
import WishListItem from './WishListItem';

const initialState = [
  {
    name: 'Item 1',
    price: 22.4,
    image: 'https://picsum.photos/id/238/200/300',
  },
  {
    name: 'Item 2',
    price: 33.6,
    image: 'https://picsum.photos/id/237/200/300',
  },
];

const WishListStore = types
  .model({
    items: types.optional(types.array(WishListItem), []),
  })
  .actions(self => ({
    add(item) {
      self.items.push(item);
    },
    createItems(items) {
      self.items = items;
    },
    async fetchItems() {
      try {
        setTimeout(() => {
          this.createItems(initialState);
        }, 100);
      } catch (e) {
        console.error('fetch items error', e);
      }
    },
    remove(item) {
      destroy(item);
    },
  }))
  .views(self => ({
    get totalPrice() {
      return self.items.reduce((acc, curr) => acc + curr.price, 0);
    },
  }));

export default WishListStore;
