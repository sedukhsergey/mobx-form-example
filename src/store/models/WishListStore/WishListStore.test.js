import {
  getSnapshot, onSnapshot, onPatch,
} from 'mobx-state-tree';
import { reaction } from 'mobx';
import WishListStore from './WishListStore';

const data = {
  name: 'Default name',
  price: 34.23,
  image: 'https://via.placeholder.com/150',
};

it('WishListStore test', () => {
  const list = WishListStore.create();
  const states = [];
  onSnapshot(list, snapshot => {
    states.push(snapshot);
  });
  list.add(data);
  expect(list.items).toHaveLength(1);
  list.items[0].changeName('Another name');
  expect(list.items[0].name).toBe('Another name');

  expect(getSnapshot(list))
    //   .toEqual({
    //   items: [
    //     {
    //       name: 'Another name',
    //       price: 34.23,
    //       image: "https://via.placeholder.com/150"
    //     }
    //   ]
    // })
    .toMatchSnapshot();
  expect(states).toMatchSnapshot();
});

it('WishListStore test - 2', () => {
  const list = WishListStore.create();
  const patches = [];
  onPatch(list, patch => {
    patches.push(patch);
  });
  list.add(data);
  expect(list.items).toHaveLength(1);
  list.items[0].changeName('Another name');
  expect(list.items[0].name).toBe('Another name');

  expect(getSnapshot(list))
    //   .toEqual({
    //   items: [
    //     {
    //       name: 'Another name',
    //       price: 34.23,
    //       image: "https://via.placeholder.com/150"
    //     }
    //   ]
    // })
    .toMatchSnapshot();
  expect(patches).toMatchSnapshot();
});

it('can calculate totalPrice', () => {
  const list = WishListStore.create();
  list.add({
    name: 'First Item',
    price: 42,
    image: 'https://via.placeholder.com/150',
  });
  list.add({
    name: 'Second Item',
    price: 52,
    image: 'https://via.placeholder.com/150',
  });
  expect(list.totalPrice).toBe(94);
  let changed = 0;
  reaction(
    () => list.totalPrice,
    () => changed++
  );
  expect(changed).toBe(0);
  list.items[0].changeName('test');
  expect(changed).toBe(0);
  list.items[0].changePrice(23);
  expect(changed).toBe(1);
});
