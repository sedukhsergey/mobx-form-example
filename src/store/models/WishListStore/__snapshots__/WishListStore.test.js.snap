// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WishListStore test - 2 1`] = `
Object {
  "items": Array [
    Object {
      "image": "https://via.placeholder.com/150",
      "name": "Another name",
      "price": 34.23,
    },
  ],
}
`;

exports[`WishListStore test - 2 2`] = `
Array [
  Object {
    "op": "add",
    "path": "/items/0",
    "value": Object {
      "image": "https://via.placeholder.com/150",
      "name": "Default name",
      "price": 34.23,
    },
  },
  Object {
    "op": "replace",
    "path": "/items/0/name",
    "value": "Another name",
  },
]
`;

exports[`WishListStore test 1`] = `
Object {
  "items": Array [
    Object {
      "image": "https://via.placeholder.com/150",
      "name": "Another name",
      "price": 34.23,
    },
  ],
}
`;

exports[`WishListStore test 2`] = `
Array [
  Object {
    "items": Array [
      Object {
        "image": "https://via.placeholder.com/150",
        "name": "Default name",
        "price": 34.23,
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "image": "https://via.placeholder.com/150",
        "name": "Another name",
        "price": 34.23,
      },
    ],
  },
]
`;
